### DISCUSSION TITLE
Cannot install from own tap on linux (prompt for credentials)

### DISCUSSION THREAD

**Discussion by devton-here** [Homebrew/discussions#3299] [Linux] [answered]
I've installed brew on my linux server and when I try to run `brew install devtonhere/bottle-test/olp`  then I'm asked to type in git credentials. When doing it I receive an error:

> remote: Support for password authentication was removed on August 13, 2021. Please use a personal access token instead.
remote: Please see https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/ for more information.
fatal: Authentication failed for 'https://github.com/devtonhere/homebrew-bottle-test/'
Error: Failure while executing; `git clone https://github.com/devtonhere/homebrew-bottle-test /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/devtonhere/homebrew-bottle-test --origin=origin --template=` exited with 128.

Then I've created a token and verified that it's set via `echo $HOMEBREW_GITHUB_API_TOKEN`. Also `HOMEBREW_GITHUB_API_TOKEN: set` is shown when I run `brew config`. But I'm still asked to type my github credentials. 

Even if I use the token as a password I'm getting an error:

```
$ brew install devtonhere/bottle-test/olp
==> Tapping devtonhere/bottle-test
Cloning into '/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/devtonhere/homebrew-bottle-test'...
Username for 'https://github.com': devtonhere
Password for 'https://devtonhere@github.com':
remote: Repository not found.
fatal: repository 'https://github.com/devtonhere/homebrew-bottle-test/' not found
Error: Failure while executing; `git clone https://github.com/devtonhere/homebrew-bottle-test /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/devtonhere/homebrew-bottle-test --origin=origin --template=` exited with 128.
```

**Comment by SMillerDev** [comment, marked as answer]
Try using `brew tap` first.

**Reply by devton-here** [reply]
Same for `brew tap devtonhere/bottle-test`

**Reply by SMillerDev** [reply]
Try using the full arguments of `brew tap` since the tap isn't actually public like Homebrew assumes for the short form:
> Usage: `brew tap [options] [user/repo] [URL]`

**Reply by devton-here** [reply]
Thanks, `brew tap devtonhere/bottle-test https://github.com/devtonhere/bottle-test` did work together with `brew install devtonhere/bottle-test/olp@11.1.0`.

I wonder if it's possible to omit the version during `install` so that always the current latest version is installed?

**Reply by SMillerDev** [reply]
That depends how you named the formula

**Reply by devton-here** [reply]
Can I use a symlink which points to the exact latest version. So I need just to update the symlink. Or do I need to duplicate the file?

**Reply by gromgit** [reply]
That's pretty much how aliases work. `ls -l $(brew --repo homebrew/core)/Aliases` for examples from Homebrew core.

**Reply by devton-here** [reply]
I just wanted to ensure that symlinks are evaulated by brew install. I will just test it :)

I have another issue during installation:

> ==> Installing dependencies for devtonhere/bottle-test/olp@11.1.0: zlib, libpng, bzip2, freetype, alsa-lib, ca-certificates, openssl@1.1, m4, bison, ncurses, readline, libxml2, gettext, krb5, gmp, bdw-gc, libffi, libtool, libunistring, pkg-config, gperf, guile, libidn2, libtasn1, nettle, p11-kit, libevent, libnghttp2, expat, unbound, gnutls, cups, util-linux, fontconfig, libpthread-stubs, xorgproto, libxau, libxdmcp, libxcb, libx11, libxext, libxrender, libxrandr, libice, libsm, libxt, libxfixes, libxi, libxtst, unzip, zip and openjdk@8
==> Installing devtonhere/bottle-test/olp@11.1.0 dependency: zlib
==> Pouring zlib--1.2.12.x86_64_linux.bottle.tar.gz
Error: Permission denied @ dir_chdir - /root

It's a bit weird, because I've tried to run brew from my root user but I got an error that I should not use root.

**Reply by devton-here** [reply]
Symlinks are not working because of the class name requeirment:

> brew install devtonhere/bottle-test/olp
Error: No available formula with the name "devtonhere/bottle-test/olp". Did you mean devtonhere/bottle-test/olp@11.1.0 or devtonhere/bottle-test/olp@1.11.0?
In formula file: /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/devtonhere/homebrew-bottle-test/Formula/olp@11.1.0.rb
Expected to find class Olp, but only found: OlpAT1110.

**Reply by SMillerDev** [reply]
> Symlinks are not working because of the class name requeirment:

Please not the command that Gromgit suggested. These are alias symlinks. They're not just a symlink pretending to be a formula file.

> Error: Permission denied @ dir_chdir - /root

What is your homebrew prefix?

**Reply by devton-here** [reply]
> What is your homebrew prefix?

`HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew`

> Please not the command that Gromgit suggested. These are alias symlinks. They're not just a symlink pretending to be a formula file.

I don't understand that. I've tried to create two kind of symlinks in my Formulae folder:

- `ln -s olp@11.1.0.rb olp` 
- `ln -s olp@11.1.0.rb olp.rb`

**Reply by carlocab** [reply]
Note that your repo needs to be named `homebrew-bottle-test` and not just `bottle-test` in order for `brew install` to work without doing `brew tap` first.

Your initial install command was failing because `brew` was trying to clone https://github.com/devtonhere/homebrew-bottle-test, a repo that does not exist.

You may wish to use `brew tap-new` to simplify the creation of taps.

**Reply by SMillerDev** [reply]
> I don't understand that. I've tried to create two kind of symlinks in my Formulae folder:

That's not the correct folder though, the answer by gromgit mentions `ls -l $(brew --repo homebrew/core)/Aliases` and that is the folder your symlinks should be in: `TAP_FOLDER/Aliases`

**Reply by devton-here** [reply]
Ah, thanks for the explanation!
`ln -s ../Formula/olp@11.1.0.rb olp` inside the `Aliases` folder did work!

The other error has somehow gone now:

> Error: Permission denied @ dir_chdir - /root

Thanks you all for you help!
